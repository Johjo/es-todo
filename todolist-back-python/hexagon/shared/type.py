from typing import NewType
from uuid import UUID

TaskKey = NewType('TaskKey', UUID)
TaskName = NewType('TaskName', str)
TaskOpen = NewType('TaskOpen', bool)

TodolistName = NewType('TodolistName', str)
TodolistContext = NewType('TodolistContext', str)
TodolistContextCount = NewType('TodolistContextCount', int)