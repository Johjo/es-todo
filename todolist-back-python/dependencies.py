from dataclasses import dataclass, field, replace
from enum import Enum
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)



    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        assert (resource_type, resource) in self.factory, f"{resource_type.value} for {resource} must be injected first"
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

