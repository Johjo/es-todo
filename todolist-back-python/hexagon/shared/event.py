from dataclasses import dataclass
from datetime import date
from uuid import UUID




@dataclass(frozen=True, eq=True)
class Event:
    pass


@dataclass(frozen=True, eq=True)
class TaskPostponedEvent(Event):
    task_key: UUID
    execution_date: date
