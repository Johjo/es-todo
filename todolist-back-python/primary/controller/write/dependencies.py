from abc import ABC
from collections import namedtuple
from dataclasses import dataclass, replace, field
from typing import Any

from hexagon.todolist.aggregate import TodolistSetPort
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.open_task import OpenTask


@dataclass(frozen=True)
class Dependencies(ABC):
    use_case_factory: dict[Any, Any] = field(default_factory=dict)
    adapter_factory: Any = None

    def get_adapter(self, port) -> Any:
        assert self.adapter_factory, f"adapter for {port} must be injected first"
        return self.adapter_factory(self)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return replace(self, use_case_factory={**self.use_case_factory, use_case: use_case_factory})


    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return replace(self, adapter_factory=adapter_factory)

    def get_use_case(self, use_case: Any) -> Any:
        assert use_case in self.use_case_factory, f"use_case for {use_case} must be injected first"
        return self.use_case_factory[use_case](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    factories = {
        TodolistCreate: todolist_create_factory,
        OpenTask : open_task_use_case_factory,
    }

    for use_case, factory in factories.items():
        dependencies = dependencies.feed_use_case(use_case, factory)
    return dependencies


def todolist_create_factory(dependencies: Dependencies):
    return TodolistCreate(dependencies.get_adapter(TodolistSetPort))

def open_task_use_case_factory(dependencies: Dependencies):
    return OpenTask(dependencies.get_adapter(TodolistSetPort))
