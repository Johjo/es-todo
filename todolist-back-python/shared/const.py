USER_KEY="user_key"