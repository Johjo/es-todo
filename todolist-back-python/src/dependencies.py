from dataclasses import dataclass, field, replace
from enum import Enum
from math import factorial
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"
    data = "data"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def feed_data(self, data_name: str, value: Any):
        return self._feed(resource_type=ResourceType.data, resource=data_name, factory=lambda _: value)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)

    def get_data(self, data_name) -> Any:
        return self._get_resource(resource_type=ResourceType.data, resource=data_name)

    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        if (resource_type, resource) not in self.factory:
            raise Exception(f"{resource_type.value} for {resource} must be injected first")
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

    def describe(self) -> str:
        def format_item(key):
            (resource_type, resource_name) = key
            resource = self._get_resource(resource_type, resource_name)
            return f"{resource_name}: {resource_type} => {str(type(resource))}"

        return "\n".join([format_item(key) for key in self.factory.keys()])
