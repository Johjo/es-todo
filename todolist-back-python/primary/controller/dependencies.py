from functools import reduce
from typing import Any

from dependencies import Dependencies
from hexagon.fvp.read.which_task import WhichTaskQuery
from hexagon.fvp.write.cancel_priority import CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    use_cases_with_factory: list[Any] = [ResetFvpSession, CancelPriority, ChooseAndIgnoreTaskFvp, TodolistCreate,
                                         OpenTaskUseCase, CloseTask, RewordTask, ImportManyTask, WhichTaskQuery]

    def feed_use_case(dep: Dependencies, use_case: Any) -> Dependencies:
        return dep.feed_use_case(use_case=use_case, use_case_factory=use_case.factory)

    return reduce(feed_use_case, use_cases_with_factory, dependencies)


