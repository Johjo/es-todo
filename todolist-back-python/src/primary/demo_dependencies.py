import os
from pathlib import Path

from dotenv import load_dotenv

from src.dependencies import Dependencies
from src.hexagon.fvp.aggregate import FvpSessionSetPort
from src.hexagon.todolist.port import TaskKeyGeneratorPort
from src.infra.memory import Memory
from src.primary.adapter_in_memory_dependencies import inject_adapter_in_memory
from src.primary.controller.read.final_version_perfected import CalendarPort
from src.primary.controller.use_case_dependencies import inject_use_cases
from src.primary.controller.write.todolist import DateTimeProviderPort
from src.primary.infrastructure_in_memory_dependencies import inject_infrastructure_in_memory
from src.secondary.calendar import Calendar
from src.secondary.datetime_provider import DateTimeProvider
from src.secondary.fvp.write.fvp_session_set_in_memory import FvpSessionSetInMemory
from src.infra.fvp_memory import FvpMemory
from src.secondary.todolist.task_key_generator_random import TaskKeyGeneratorRandom

memory = Memory()
fvp_memory = FvpMemory()

def inject_all_dependencies(dependencies: Dependencies) -> Dependencies:
    dependencies = inject_use_cases(dependencies)
    dependencies = inject_adapter_in_memory(dependencies)
    dependencies = inject_infrastructure_in_memory(dependencies=dependencies, memory=memory, fvp_memory=fvp_memory)

    dependencies = dependencies.feed_adapter(TaskKeyGeneratorPort, lambda _: TaskKeyGeneratorRandom())
    dependencies = dependencies.feed_adapter(DateTimeProviderPort, DateTimeProvider.factory)
    dependencies = dependencies.feed_adapter(CalendarPort, Calendar.factory)
    load_dotenv()
    dependencies = dependencies.feed_path("static_path", lambda _: Path(os.environ["STATIC_PATH"]))
    dependencies = dependencies.feed_adapter(FvpSessionSetPort, FvpSessionSetInMemory.factory)

    return dependencies


