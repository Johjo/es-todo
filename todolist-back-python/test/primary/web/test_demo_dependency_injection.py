import pytest
from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter  # type: ignore

from src.dependencies import Dependencies
from src.primary.demo_dependencies import inject_all_dependencies
from src.shared.const import USER_KEY


@pytest.fixture
def sut() -> Dependencies:
    dependencies = Dependencies.create_empty()
    dependencies = inject_all_dependencies(dependencies)
    dependencies = dependencies.feed_data(data_name=USER_KEY, value="test")
    return dependencies


def test_describe_dependencies(sut: Dependencies) -> None:
    verify(sut.describe(), reporter=PythonNativeReporter())
