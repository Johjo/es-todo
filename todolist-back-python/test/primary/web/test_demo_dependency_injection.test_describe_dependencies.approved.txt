<class 'src.hexagon.fvp.write.reset_fvp_session.ResetFvpSession'>: ResourceType.use_case => <class 'src.hexagon.fvp.write.reset_fvp_session.ResetFvpSession'>
<class 'src.hexagon.fvp.write.cancel_priority.CancelPriority'>: ResourceType.use_case => <class 'src.hexagon.fvp.write.cancel_priority.CancelPriority'>
<class 'src.hexagon.fvp.write.choose_and_ignore_task.ChooseAndIgnoreTaskFvp'>: ResourceType.use_case => <class 'src.hexagon.fvp.write.choose_and_ignore_task.ChooseAndIgnoreTaskFvp'>
<class 'src.hexagon.todolist.write.create_todolist.TodolistCreate'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.create_todolist.TodolistCreate'>
<class 'src.hexagon.todolist.write.open_task.OpenTaskUseCase'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.open_task.OpenTaskUseCase'>
<class 'src.hexagon.todolist.write.close_task.CloseTask'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.close_task.CloseTask'>
<class 'src.hexagon.todolist.write.reword_task.RewordTask'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.reword_task.RewordTask'>
<class 'src.hexagon.todolist.write.import_many_task.ImportManyTask'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.import_many_task.ImportManyTask'>
<class 'src.hexagon.fvp.read.which_task.WhichTaskQuery'>: ResourceType.use_case => <class 'src.hexagon.fvp.read.which_task.WhichTaskQuery'>
<class 'src.hexagon.todolist.write.postpone_task.PostPoneTask'>: ResourceType.use_case => <class 'src.hexagon.todolist.write.postpone_task.PostPoneTask'>
<class 'src.hexagon.todolist.port.TodolistSetPort'>: ResourceType.adapter => <class 'src.secondary.todolist.todolist_set.todolist_set_in_memory.TodolistSetInMemory'>
<class 'src.hexagon.fvp.read.which_task.TodolistPort'>: ResourceType.adapter => <class 'src.secondary.fvp.read.which_task.todolist_memory.TodolistInMemory'>
<class 'src.primary.controller.read.todolist.TodolistSetReadPort'>: ResourceType.adapter => <class 'src.secondary.todolist.todolist_set_read.todolist_set_read_memory.TodolistSetReadInMemory'>
<class 'src.hexagon.fvp.aggregate.FvpSessionSetPort'>: ResourceType.adapter => <class 'src.secondary.fvp.write.fvp_session_set_in_memory.FvpSessionSetInMemory'>
<class 'src.infra.memory.Memory'>: ResourceType.infrastructure => <class 'src.infra.memory.Memory'>
<class 'src.infra.fvp_memory.FvpMemory'>: ResourceType.infrastructure => <class 'src.infra.fvp_memory.FvpMemory'>
<class 'src.hexagon.todolist.port.TaskKeyGeneratorPort'>: ResourceType.adapter => <class 'src.secondary.todolist.task_key_generator_random.TaskKeyGeneratorRandom'>
<class 'src.primary.controller.write.todolist.DateTimeProviderPort'>: ResourceType.adapter => <class 'src.secondary.datetime_provider.DateTimeProvider'>
<class 'src.primary.controller.read.final_version_perfected.CalendarPort'>: ResourceType.adapter => <class 'src.secondary.calendar.Calendar'>
static_path: ResourceType.path => <class 'pathlib.WindowsPath'>
user_key: ResourceType.data => <class 'str'>
